az-hadoop-jobtype-plugin/src/main/java/azkaban/jobtype/HadoopSecureWrapperUtils.java

```diff
@@ -59,32 +66,61 @@ private static UserGroupInformation createSecurityEnabledProxyUser(String userTo
         + " to " + fileLocation);
     System.setProperty(HadoopSecurityManager.MAPREDUCE_JOB_CREDENTIALS_BINARY, fileLocation);
 
-    UserGroupInformation loginUser = null;
-
-    loginUser = UserGroupInformation.getLoginUser();
-    log.info("Current logged in user is " + loginUser.getUserName());
+    // log the tokens of getLoginUser() and monitor the copying
+    UserGroupInformation loginUser = UserGroupInformation.getLoginUser();
+    log.info("Current logged in user is " + loginUser);
 
     UserGroupInformation proxyUser = UserGroupInformation.createProxyUser(userToProxy, loginUser);
 
+    log.info(String.format("Copy from loginUser [%s] to proxyUser [%s]", loginUser, proxyUser));
     for (Token<?> token : loginUser.getTokens()) {
       proxyUser.addToken(token);
+      log.info(String.format("Token = %s, %s, %s ", token.getKind(), token.getService(),
+          Arrays.toString(token.getIdentifier())));
     }
     proxyUser.addCredentials(loginUser.getCredentials());
+
+    // read tokens from the file and put into proxyUser
+    if (hadoopSecurityManager != null) {
+      Credentials creds = hadoopSecurityManager.getTokens(new File(fileLocation), log);
+      log.info(String.format("Loading tokens from file [%s] to proxyUser [%s]", fileLocation,
+          proxyUser));
+      for (Token<?> token : creds.getAllTokens()) {
+        proxyUser.addToken(token);
+        log.info(String.format("Token = %s, %s, %s ", token.getKind(), token.getService(),
+            Arrays.toString(token.getIdentifier())));
+      }
+      proxyUser.addCredentials(creds);
+    }
+
+    log.info("token copy finished for " + loginUser.getUserName());
```

Comment:
The login user is `azdata`. proxy user is any proxy user P. wouldn't it be sufficient if we do this alone? I am trying to figure out why we are doing the copy from line 76 to line 81. am i missing something?

Reply:
I don't think using any user out of nowhere is good, plus this is the intend of `proxy`User, right?
Duplicate code but these are 2 sources of tokens, first is copy from login user, second is from token-file.